[Index management] Data stream edit data retention

sabarasaba opened the pull request:
Fixes: https://github.com/elastic/kibana/issues/165136

## Summary
This PR adds support for editing data retention for a data stream.


##### How to test
1. Start elasticsearch with `yarn es snapshot --license=trial` and serverless search solution with `yarn start`
2. Navigate to console and execute the commands listed below in order to create a policy
3. Navigate to `index management` and click on `data streams` tab
4. Verify that:
* You can edit the data retention, and its reflected in the table and details flyout
* If you set infinite data retention, it should be reflected in the table and details flyout. And when the edit form is open again the toggle should be enabled.
* If infinite toggle is enabled, you shouldnt be allowed to edit data retention values
* If the data stream is managed by ILM, the edit data retention option shouldnt be shown

<details>
<summary>Creating a test data stream</summary>

```
PUT _index_template/b-index-template
{
  "index_patterns": ["b-data-stream*"],
  "data_stream": { },
  "priority": 500,
  "template": {
    "lifecycle": {
      "enabled": false
    }
  },
  "_meta": {
    "description": "Template with data stream lifecycle"
  }
}

PUT _data_stream/b-data-stream

PUT b-data-stream/_bulk
{ "create":{ } }
{ "@timestamp": "2099-05-06T16:21:15.000Z", "message": "192.0.2.42 - - [06/May/2099:16:21:15 +0000] \"GET /images/bg.jpg HTTP/1.0\" 200 24736" }
{ "create":{ } }
{ "@timestamp": "2099-05-06T16:25:42.000Z", "message": "192.0.2.255 - - [06/May/2099:16:25:42 +0000] \"GET /favicon.ico HTTP/1.0\" 200 3638" }
```

</details>


<details>
<summary>Screenshots</summary>

![Screenshot 2023-09-29 at 11 15 50](https://github.com/elastic/kibana/assets/1191206/4c19ac4a-ea5a-4e9a-978f-b1eb29516a97)


</details>

sabarasaba commented:
@elasticmachine merge upstream

kibanamachine commented:
merge conflict between base and head

sabarasaba commented on x-pack/plugins/index_management/public/application/sections/home/data_stream_list/edit_data_retention_modal/edit_data_retention_modal.tsx:
```diff
@@ -0,0 +1,323 @@
+  onClose: (data?: { hasUpdatedDataRetention: boolean }) => void;
+}
+
+const DEFAULT_DATA_RETENTION_PERIOD = '7d';
```
I'm defaulting in the UI  to `7d` retention in case there arent any values set in the dataRetention field

elasticmachine commented:
Pinging @elastic/platform-deployment-management (Team:Deployment Management)

alisonelizabeth commented:
@elasticmachine merge upstream

alisonelizabeth commented on x-pack/test/api_integration/apis/management/index_management/data_streams.ts:
```diff
@@ -210,6 +210,25 @@ export default function ({ getService }: FtrProviderContext) {
+        expect(body).to.eql({ success: true });
+      });
+    });
+
```
Is it worth adding another test for the infinite case (ie `{}`)

alisonelizabeth commented on x-pack/test/api_integration/apis/management/index_management/data_streams.ts:
```diff
@@ -210,6 +210,25 @@ export default function ({ getService }: FtrProviderContext) {
       });
     });
 
+    describe('Update', () => {
```
Can we add tests for serverless as well 🙏

sabarasaba replied:
I've added functional tests for serverless, but we might get hit by the same problem the other index_management functional tests have now that make them flaky! Hopefully those will be sorted out soon 😀

alisonelizabeth commented on x-pack/plugins/index_management/common/lib/data_stream_serialization.ts:
```diff
@@ -54,3 +54,19 @@ export function deserializeDataStreamList(
   return dataStreamsFromEs.map((dataStream) => deserializeDataStream(dataStream));
 }
+
+export const splitSizeAndUnits = (field: string): { size: string; unit: string } => {
```
I sort of recall having a helper/lib for this already but I can't seem to find it

sabarasaba replied:
We have it in ILM, but ILM is only present in stateless so we cant require it from the plugin. I decided to copy it over and added some extra tests at least for now

alisonelizabeth reviewed (commented):
Really great work @sabarasaba! Thanks for the quick turnaround on this. Left some comments in the code.

Also, a few other observations:

- I noticed the bug you mentioned to me about "Indices" sometimes showing twice in the flyout. This indeed appears to be a regression of https://github.com/elastic/kibana/pull/165263. I think in `data_stream_detail_panel.tsx`, if you change L79 to as follows it should fix it:
```
const descriptionListColumnTwo = descriptionListItems.slice(midpoint, descriptionListItems.length);
```

- [Follow up] If a user selects "Never delete data" and saves, I think we should represent this in the table view as well (right now this translates as an empty table cell for the "Data retention" column). This is probably more appropriate as a follow-up PR, but just something I noticed and wanted to call out.
- [Follow up] We might want to consider creating human readable units to display in the table. For example `7m` could be interpreted as "minutes" or "months". Again, not directly related to your work, just something I noticed.

florent-leborgne commented:
Hey @sabarasaba, I'll review the copy today on behalf of @gchaps.

sabarasaba commented on x-pack/plugins/index_management/__jest__/client_integration/home/data_streams_tab.helpers.ts:
```diff
@@ -176,8 +177,13 @@ export const setup = async (
   const clickDeleteDataStreamButton = () => {
-    const { find } = testBed;
-    find('deleteDataStreamButton').simulate('click');
+    testBed.find('manageDataStreamButton').simulate('click');
```
Wrapping it with act breaks the tests, I've noticed most click handlers in this testsuite dont use act so wondering if theres something else going on underneath. I think I'll leave it as is for now 🤔

sabarasaba commented on x-pack/plugins/index_management/public/application/sections/home/data_stream_list/data_stream_detail_panel/data_stream_detail_panel.tsx:
```diff
@@ -324,20 +404,22 @@ export const DataStreamDetailPanel: React.FunctionComponent<Props> = ({
               </EuiButtonEmpty>
             </EuiFlexItem>
 
-            {!isLoading && !error && dataStream?.privileges.delete_index ? (
```
There is currently a CIT that tests for permissions and verifies this logic remains! So I think we are safe here 👯

sabarasaba commented:
@elasticmachine merge upstream

sabarasaba commented:
Thanks for the review @alisonelizabeth! I've addressed your feedback with https://github.com/elastic/kibana/pull/167006/commits/59ba5f6c3a0989351c60feeb5171eabdbef51d6d and created [a new issue](https://github.com/elastic/kibana/issues/167475) for dealing with the two issues you mentioned above about being more explicit about data retention period in the UI.

florent-leborgne commented on x-pack/plugins/index_management/public/application/sections/home/data_stream_list/edit_data_retention_modal/edit_data_retention_modal.tsx:
```diff
@@ -0,0 +1,327 @@
+                <p>
+                  <FormattedMessage
+                    id="xpack.idxMgmt.dataStreamsDetailsPanel.editDataRetentionModal.configuredByILMWarningText"
+                    defaultMessage="Since this data stream is configured by ILM, you must also update the ILM policy to match the new data retention period."
```
```suggestion
                    defaultMessage="The data retention set for this data stream currently overrides the retention settings of the Index Lifecycle Policy attached to this data stream. Update the corresponding policy to match the retention period you need."
```

I'm not clear on the behavior implied by this message, so not sure what to suggest here for the title and description. So feel free to ignore or correct my suggestion here.
Do settings conflict? What happens or which setting counts if values are different here and in the policy?

sabarasaba replied:
I think we might leave this as is now, as we are going to update it in the upcoming days to address a bunch of scenarios that could happen here. See https://github.com/elastic/kibana/issues/167414

florent-leborgne replied:
Sounds good, let me know when I can help :)

florent-leborgne commented on x-pack/plugins/index_management/public/application/sections/home/data_stream_list/edit_data_retention_modal/edit_data_retention_modal.tsx:
```diff
@@ -0,0 +1,327 @@
+                title={i18n.translate(
+                  'xpack.idxMgmt.dataStreamsDetailsPanel.editDataRetentionModal.configuredByILMWarningTitle',
+                  {
+                    defaultMessage: 'Override ILM data retention',
```
```suggestion
                    defaultMessage: 'Current data retention conflicts with ILM policy ',
```
Feel free to ignore or adjust, see my comment below on the callout body

florent-leborgne reviewed (commented):
Hey @sabarasaba. I've suggested a few edits. Also have a few questions/remarks for things I'm not sure of:

- I was not sure of what to suggest for the warning, but I think we can make it clearer. See my comments in the review

- Another thing that caught my attention are the shortened time values we use. 
     - I understand that they match with the API but "nanos" "micros" etc. sound quite unusual to me compared to common abbreviations (d, h, min, sec, ms, µs, ns) or fully developed forms if we have enough space for it, like `3000 microseconds`
     - This maybe also has to do with the API but we render the value with no space in the flyout, while the best practice would be to have a (non-breaking) space between the number and the unit. (30 ns, 4 days, instead of 30ns, 4days). I see it's similar for the storage size so at least it's consistent, but if we can consider changing this at some point in the UI, it would be great.
    
<img width="120" alt="image" src="https://github.com/elastic/kibana/assets/10208282/4ce86954-1603-494e-901c-64ce7580f7c8">

- Tooltips for "Data retention" differ in the flyout and in the data stream list. I'm not sure if they are still correct, especially the long one? Happy to help revisit them once I'm a bit more sure about the actual behavior ^^'
<img width="285" alt="Pasted Graphic" src="https://github.com/elastic/kibana/assets/10208282/2f729010-e34a-4f6e-925d-efecfe5669b5">
<img width="275" alt="Pasted Graphic 1" src="https://github.com/elastic/kibana/assets/10208282/b0c25f29-1b73-4a16-b890-1a240e9f7ef9">

- Not a content comment, but is it the expected behavior that once we edit the data retention, the flyout also closes?

sabarasaba commented on x-pack/plugins/index_management/public/application/sections/home/data_stream_list/edit_data_retention_modal/edit_data_retention_modal.tsx:
```diff
@@ -0,0 +1,327 @@
+    label: i18n.translate(
+      'xpack.idxMgmt.dataStreamsDetailsPanel.editDataRetentionModal.infiniteRetentionPeriodField',
+      {
+        defaultMessage: 'Never delete data',
```
Thanks for the suggestion! Not sure if we want to add more things here.. we will revisti most of this copy in a subsequent PR with https://github.com/elastic/kibana/issues/167414

florent-leborgne replied:
Sure thanks, we can keep it to the simpler label suggestion for now :)

sabarasaba commented:
Thanks for the review @florent-leborgne! I've addressed some of your feedback with https://github.com/elastic/kibana/pull/167006/commits/003433641d7fee53b1b5c6072abc219f1c736998 and I've created [this issue](https://github.com/elastic/kibana/issues/167475) to address some of the other smaller things you mentioned.

In regards of the tooltip content, I'm not sure if thats a regresion or not cc: @alisonelizabeth.

florent-leborgne reviewed (approved):
Approving latest copy while noting that pending comments will be addressed through separate issues as discussed. Thanks!

yuliacech commented on x-pack/plugins/index_management/__jest__/client_integration/helpers/test_subjects.ts:
```diff
@@ -77,6 +77,13 @@ export type TestSubjects =
   | 'createIndexCancelButton'
+  | 'manageDataStreamButton'
+  | 'dataRetentionValue'
+  | 'configuredByILMWarning'
```
nit: this test subject is probably not needed anymore

yuliacech commented on x-pack/plugins/index_management/__jest__/client_integration/home/data_streams_tab.test.ts:
```diff
@@ -423,6 +491,33 @@ describe('Data Streams tab', () => {
       expect(findDetailPanelIlmPolicyLink().prop('href')).toBe('/test/my_ilm_policy');
     });
 
+    test('with ILM updating data retention should be dissabled', async () => {
```
typo
```suggestion
    test('with ILM updating data retention should be disabled', async () => {
```

yuliacech commented on x-pack/plugins/index_management/public/application/sections/home/data_stream_list/edit_data_retention_modal/edit_data_retention_modal.tsx:
```diff
@@ -0,0 +1,327 @@
+              message: i18n.translate(
+                'xpack.idxMgmt.dataStreamsDetailsPanel.editDataRetentionModal.dataRetentionFieldRequiredError',
+                {
+                  defaultMessage: 'A data retention value is required.',
```
nit: I think for `value === 0`, I would display a separate message. Maybe even put it together with `value < 0`, something like "a positive value is required".

yuliacech reviewed (approved):
Thanks a lot for working on this feature, @sabarasaba! 
Left just a couple of nits in the code, but the latest changes LGTM 👍

amyjtechwriter reviewed (commented):
Added link LGTM

sabarasaba commented:
@elasticmachine merge upstream
